src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -22,6 +22,23 @@
 from pants.rules.core.strip_source_root import SourceRootStrippedSources
 
 
+def get_timeout_seconds_for_target(test_target_timeout, timeout_default, timeout_maximum):
```

Comment:
type hints appreciated

Reply:
Also, I recommend using keyword only args like this:

```python
def get_timeout_seconds_for_target(*, test_target_timeout, timeout_default, timeout_maximum):
```

These arguments are all similar and very easy to get mixed up. They also have the same type signatures (or at least some of them do), so MyPy wouldn't catch passing the wrong arg.

Reply:
Finally, consider naming this `calculate_timeout_seconds`. The keyword arg `test_target_timeout` expresses that this is specific to a timeout.

Reply:
One more thing about the logic of this function. I think that we allow for `timeout_default` to be `Optional`. So, there's a case where every single argument is `None`. In that case, I think that the function should return `None`.

Then, in the callsite in the rule, add something like `get_timeout_seconds_for_target(...) or 9999` because we must provide a `timeout` to the EPR.